### PR TITLE
remove `HierarchyNotRespected` validation error

### LINKED ISSUE #638: Remove `HierarchyNotRespected` validation error
### Describe the improvement you'd like to request

According to [this comment](https://github.com/cedar-policy/cedar/blob/5f62c6df06b59abc5634d6668198a826839c6fb7/cedar-policy-validator/src/typecheck.rs#L1871) we return the `HierarchyNotRespected` error in some cases where we could soundly type an expression as `False`. This error was introduced (IIRC) to avoid having to update the Dafny type soundness proof.

Now that we've switched to using Lean, and tweaked the validation [property being tested](https://github.com/cedar-policy/cedar-spec/pull/210), I wonder if we could remove this error variant.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### Is this something that you'd be interested in working on?

- [ ] 👋 I may be able to implement this internal improvement
- [ ] ⚠️ This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by cdisselkoen**
## Description of changes

As suggested in #638, removes the `HierarchyNotRespected` validation error (well, except from the public error enum, where we leave it to avoid a breaking change, but adjust documentation to indicate it is never returned).  This simplifies the typechecker code considerably.  

I'm a little unclear from the discussion on #638 whether the Lean model needs any corresponding changes or not.

## Issue #, if available

Fixes #638 

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A bug fix or other functionality change requiring a patch to `cedar-policy`.

I confirm that this PR (choose one, and delete the other options):

- [x] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- Not sure about changes to the Lean model, see above

I confirm that [`docs.cedarpolicy.com`](https://docs.cedarpolicy.com/) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar language specification.

### CODE REVIEW

**Review by john-h-kastner-aws — APPROVED** [review]
I happy to see this changes,  though I'm not 100% sure it will pass DRT. It might be returning `Bool` when the lean impl gives a more precise `False` type. Fixing any discrepancy there should also be a backwards compatible change, but we should take a close look at what both are doing with `in`.

If it does pass, then we _might_ be able to use `AgreeOnAll` here if this is the only error the Rust reports but Lean doesn't.

https://github.com/cedar-policy/cedar-spec/blob/main/cedar-drt/src/lean_impl.rs#L516

File: cedar-policy-validator/src/diagnostics/validation_errors.rs [hunk 1]
```diff
@@ -460,28 +460,19 @@ impl Diagnostic for FunctionArgumentValidation {
 
 /// Structure containing details about a hierarchy not respected error
 #[derive(Debug, Clone, Hash, Eq, PartialEq, Error)]
-#[error("for policy `{policy_id}`, operands to `in` do not respect the entity hierarchy")]
+#[error("Internal invariant violated: `HierarchyNotRespected` error should never occur. Please file an issue")]
 pub struct HierarchyNotRespected {
     /// Source location
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy-validator/src/diagnostics/validation_errors.rs:465]: `wrap_core_error!` in `api/err/validation_errors.rs` needs them.  Taking a couple steps back, since the variant still exists in the public `ValidationError`, we need to support `ValidationError::policy_id()` on it.  And you could say we could return a fictitious policy id or something, but if this error actually ever does get thrown in violation of the internal invariant (due to some future refactor), we'll be glad to have the policy id and source location

File: cedar-policy-validator/src/typecheck.rs [hunk 2]
```diff
@@ -1992,85 +1990,21 @@ impl<'a> Typechecker<'a> {
                     ),
                 }
                 .then_typecheck(|type_of_in, _| {
-                    if !self.mode.is_strict() {
-                        TypecheckAnswer::success(type_of_in)
-                    } else if matches!(type_of_in.data(), Some(Type::False)) {
+                    if self.mode.is_strict() && matches!(type_of_in.data(), Some(Type::False)) {
                         TypecheckAnswer::success(
                             ExprBuilder::with_data(Some(Type::False)).val(false),
```
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy-validator/src/typecheck.rs:1995]: I'll do that in a separate PR since this one is a little delicate as-is and the change you suggest is technically orthogonal

**Comment by cdisselkoen** [comment]
> It might be returning `Bool` when the lean impl gives a more precise `False` type.

I looked at this Rust code some, and I think/hope that the logic I removed was already performed in `type_of_var_in_entity_literals()` and friends.  The logic I removed only ever results in `HierarchyNotRespected` or `success(type_of_in)`; it never returns `False` when `type_of_in` wasn't already `False`.  So removing this logic shouldn't stop us from returning `False` in any situation where we were already returning `False` before this PR.

**Comment by john-h-kastner-aws** [comment]
> > It might be returning `Bool` when the lean impl gives a more precise `False` type.
> 
> I looked at this Rust code some, and I think/hope that the logic I removed was already performed in `type_of_var_in_entity_literals()` and friends. The logic I removed only ever results in `HierarchyNotRespected` or `success(type_of_in)`; it never returns `False` when `type_of_in` wasn't already `False`. So removing this logic shouldn't stop us from returning `False` in any situation where we were already returning `False` before this PR.

IIUC, the DRT currently passes if Rust returns an errors when lean doesn't. So, if rust has stopped returning errors here, there could be a precision difference in the non-error response.

**Review by john-h-kastner-aws — APPROVED** [review]
This looks good to me. The concerns I mention will be fully addressed just by running the diff tests with this change merged, so we just need another review to look this over.

**Comment by cdisselkoen** [comment]
Corpus tests are failing in CI, probably because there are edge cases where this PR changes the validation behavior from fail -> pass (as noted in the changelog entry).  Adjusting the corpus test expectation will have to be a different PR against cedar-integration-tests

**Comment by cdisselkoen** [comment]
note on the status of this: have been waiting for cedar-policy/cedar-integration-tests#14 to be merged first, to avoid conflicts in updating the corpus tests

File: cedar-policy-validator/src/typecheck.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy-validator/src/typecheck.rs:1955]: I think that change was done in #1462?

[the remark this answers is not in the document]
Comment (cdisselkoen) [reply cedar-policy-validator/src/typecheck.rs]: Updated this PR, now simplified further as you say.  (we just return `success(type_of_in)` in all cases)

**Comment by cdisselkoen** [comment]
Ready for re-review. Corpus test failure is expected, see comments above; we will need to regenerate corpus tests because of this PR, which will be a separate PR to cedar-integration-tests. Also, DRT failure is because as of this writing we have merged #1535 but not cedar-policy/cedar-spec#587.

**Review by john-h-kastner-aws — APPROVED** [review]

**Review by shaobo-he-aws — APPROVED** [review]
